File: app/local_plugins/assistant_server/ios/Classes/AssistantSession.swift
import Flutter
import Foundation
import ai_plugin_interface
import os.log

/// 一个活跃的 AI 助理会话（iOS）—— 与 Android `AssistantSession` 对齐。
///
/// 通过 BinaryMessenger 与 device_jieli 通信：
///   - 控制：`device_jieli/method` MethodChannel（assistantPortEnter / Exit / ReportPlayback）
///   - 上行：NotificationCenter `JieliAssistantPortAudio` / `JieliAssistantPortError`
///
/// 这样避免了 Pod↔SPM 的 Swift 模块导入问题。
final class AssistantSession {

    private static let log = OSLog(subsystem: "com.aiagent.assistant_server", category: "Session")
    private static let connectTimeoutSec: TimeInterval = 10

    let sessionId: String
    private let request: AssistantRequest
    private let emit: ([String: Any?]) -> Void
    private let deviceMethod: FlutterMethodChannel

    private let scope = DispatchQueue(label: "AssistantSession.\(UUID().uuidString)", qos: .userInitiated)

    private var agent: NativeAgent?
    private var sink: AssistantSinkAdapter?
    private var audioObserver: NSObjectProtocol?
    private var errorObserver: NSObjectProtocol?
    private let ttsSink: TtsSink

    private enum State { case starting, active, stopping, stopped, error }
    private let stateLock = NSLock()
    private var state: State = .starting

    init(sessionId: String, request: AssistantRequest,
         deviceMethod: FlutterMethodChannel,
         emit: @escaping ([String: Any?]) -> Void) {
        self.sessionId = sessionId
        self.request = request
        self.deviceMethod = deviceMethod
        self.emit = emit
        self.ttsSink = TtsSink(deviceMethod: deviceMethod)
    }

    /// 启动会话；失败抛 AssistantSessionError，Core 负责 emit error。
    func start() {
        scope.async { [weak self] in
            guard let self = self else { return }
            do {
                try self.startBlocking()
            } catch {
                let code: String
                if let err = error as? AssistantSessionError { code = err.code }
                else { code = "assistant.start_failed" }
                self.markError(code: code, message: "\(error)")
            }
        }
    }

    private func startBlocking() throws {
        // 1. 创建 agent 实例（agents_server 同等机制，按 agentType 工厂创建）
        let a = try NativeAgentRegistry.shared.create(request.agentType)
        self.agent = a

        // 2. SinkAdapter 把 agent 事件路由到 assistant_server EventChannel
        let s = AssistantSinkAdapter(sessionId: sessionId, emit: emit)
        self.sink = s
        a.initialize(config: request.agentConfig, eventSink: s)

        // 3. connectService（云端建链）+ 等 onAgentReady
        a.connectService()
        do {
            try s.awaitReady(timeout: Self.connectTimeoutSec)
        } catch {
            throw error
        }

        // 4. 校验 agent 外部音频能力（必须 acceptsPcm）
        let cap = a.externalAudioCapability()
        guard cap.acceptsPcm else {
            throw AssistantSessionError.agentUnsupported(message: "agent doesn't accept PCM external audio")
        }

        // 5. 注册 NotificationCenter observer —— 接 device_jieli 端口推上来的音频帧
        let nc = NotificationCenter.default
        audioObserver = nc.addObserver(forName: Notification.Name("JieliAssistantPortAudio"),
                                       object: nil, queue: nil) { [weak self] note in
            guard let self = self, let info = note.userInfo,
                  let pcm = info["pcm"] as? Data,
                  let agent = self.agent else { return }
            agent.pushExternalAudioFrame(pcm)
        }
        errorObserver = nc.addObserver(forName: Notification.Name("JieliAssistantPortError"),
                                       object: nil, queue: nil) { [weak self] note in
            guard let self = self, let info = note.userInfo else { return }
            let code = (info["code"] as? String) ?? "device.error"
            let msg  = (info["message"] as? String) ?? ""
            self.emit(AssistantEvents.error(sessionId: self.sessionId, code: code, message: msg))
        }

        // 6. 启动 agent 的外部音频通路 + TTS 反向 sink
        let format = ExternalAudioFormat.pcmS16LE16kMono20ms
        do {
            try a.startExternalAudio(format: format, sink: ttsSink)
        } catch {
            throw AssistantSessionError.agentUnsupported(message: "startExternalAudio failed: \(error)")
        }

        // 7. 进入设备端 port —— 此后耳机帧会通过 Notification 派发到我们的 observer
        let result = invokeDeviceMethodSync(
            method: "assistantPortEnter",
            args: ["sampleRate": NSNumber(value: 16000)]
        )
        if case .failure(let err) = result {
            throw AssistantSessionError.enterFailed(message: err.localizedDescription)
        }

        stateLock.lock(); state = .active; stateLock.unlock()
        emit(AssistantEvents.sessionState(sessionId: sessionId, state: "active"))
        os_log("session=%{public}@ active", log: Self.log, type: .debug, sessionId)
    }

    /// 停止会话；幂等。
    func stop() {
        stateLock.lock()
        if state == .stopped || state == .stopping {
            stateLock.unlock(); return
        }
        state = .stopping
        stateLock.unlock()

        emit(AssistantEvents.sessionState(sessionId: sessionId, state: "stopping"))

        scope.async { [weak self] in
            guard let self = self else { return }
            // 反注册 NotificationCenter
            if let o = self.audioObserver { NotificationCenter.default.removeObserver(o) }
            if let o = self.errorObserver { NotificationCenter.default.removeObserver(o) }
            self.audioObserver = nil
            self.errorObserver = nil

            // 退出 device port
            _ = self.invokeDeviceMethodSync(method: "assistantPortExit", args: [:])

            // 释放 agent
            self.agent?.stopExternalAudio()
            self.agent?.disconnectService()
            self.agent?.release()
            self.agent = nil
            self.sink = nil

            self.stateLock.lock(); self.state = .stopped; self.stateLock.unlock()
            self.emit(AssistantEvents.sessionState(sessionId: self.sessionId, state: "stopped"))
            os_log("session=%{public}@ stopped", log: Self.log, type: .debug, self.sessionId)
        }
    }

    func markError(code: String, message: String) {
        stateLock.lock()
        if state == .stopped || state == .error {
            stateLock.unlock(); return
        }
        state = .error
        stateLock.unlock()
        emit(AssistantEvents.error(sessionId: sessionId, code: code, message: message, fatal: true))
        emit(AssistantEvents.sessionState(sessionId: sessionId, state: "error", errorMessage: message))
        stop()
    }

    // MARK: - Device MethodChannel sync invocation

    @discardableResult
    private func invokeDeviceMethodSync(method: String, args: [String: Any]) -> Result<Any?, NSError> {
        let sem = DispatchSemaphore(value: 0)
        var resultBox: Result<Any?, NSError> = .success(nil)
        DispatchQueue.main.async {
            self.deviceMethod.invokeMethod(method, arguments: args) { res in
                if let err = res as? FlutterError {
                    resultBox = .failure(NSError(
                        domain: "device_jieli", code: -1,
                        userInfo: [NSLocalizedDescriptionKey: "\(err.code): \(err.message ?? "")"]
                    ))
                } else {
                    resultBox = .success(res)
                }
                sem.signal()
            }
        }
        if sem.wait(timeout: .now() + 5) == .timedOut {
            return .failure(NSError(domain: "device_jieli", code: -2,
                                    userInfo: [NSLocalizedDescriptionKey: "method \(method) timeout"]))
        }
        return resultBox
    }
}

/// AI agent 的 TTS PCM → device_jieli 端口回灌 sink。
private final class TtsSink: ExternalAudioSink {
    private weak var deviceMethod: FlutterMethodChannel?

    init(deviceMethod: FlutterMethodChannel) {
        self.deviceMethod = deviceMethod
    }

    func onTtsFrame(_ frame: ExternalAudioFrame) {
        guard frame.codec == .pcmS16LE else { return }
        let dm = deviceMethod
        // 投递到主线程（FlutterMethodChannel 要求 main）
        let args: [String: Any] = [
            "pcm": FlutterStandardTypedData(bytes: frame.bytes),
            "sampleRate": NSNumber(value: frame.sampleRate),
            "channels": NSNumber(value: frame.channels),
            "final": frame.isFinal,
        ]
        DispatchQueue.main.async {
            dm?.invokeMethod("assistantPortReportPlayback", arguments: args, result: nil)
        }
    }

    func onError(code: String, message: String) {
        // 暴露给上层的 error 由 SinkAdapter 处理；这里只记日志。
        os_log("TtsSink error: %{public}@ %{public}@", log: OSLog.default, type: .error, code, message)
    }
}

/// startAssistant 入参。
struct AssistantRequest {
    let agentType: String
    let agentConfig: NativeAgentConfig
    let userLanguage: String
}

File: apps/client/local_plugins/assistant_server/android/src/main/kotlin/com/aiagent/assistant_server/AssistantSession.kt
package com.aiagent.assistant_server

import android.content.Context
import android.util.Log
import com.aiagent.device_plugin_interface.AssistantAudioCodec
import com.aiagent.device_plugin_interface.AssistantAudioFormat
import com.aiagent.device_plugin_interface.AssistantPlaybackFrame
import com.aiagent.device_plugin_interface.DeviceAssistantPort
import com.aiagent.plugin_interface.ExternalAudioFormat
import com.aiagent.plugin_interface.ExternalAudioFrame
import com.aiagent.plugin_interface.ExternalAudioSink
import com.aiagent.plugin_interface.NativeAgent
import com.aiagent.plugin_interface.NativeAgentConfig
import com.aiagent.plugin_interface.NativeAgentRegistry
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.collect

/**
 * 一个活跃的 AI 助理会话。
 *
 * 通过 vendor-agnostic 的 [DeviceAssistantPort] 打通设备 PCM 上下行：
 *  - 上行：[DeviceAssistantPort.audioFrames] → chat agent `pushExternalAudioFrame`；
 *  - 下行：chat agent TTS PCM → [DeviceAssistantPort.reportPlayback] 回灌耳机扬声器。
 *
 * 生命周期：starting → active → stopping → stopped/error。
 */
internal class AssistantSession(
    val sessionId: String,
    private val context: Context,
    private val request: AssistantRequest,
    private val emit: (Map<String, Any?>) -> Unit,
) {
    companion object {
        private const val TAG = "AssistantSession"
        private const val CONNECT_TIMEOUT_MS = 10_000L
    }

    enum class State { STARTING, ACTIVE, STOPPING, STOPPED, ERROR }

    @Volatile var state: State = State.STARTING
        private set

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private var agent: NativeAgent? = null
    private var devicePort: DeviceAssistantPort? = null

    /** 启动整条管道；失败抛出（调用方负责 emit error）。 */
    suspend fun start() {
        // 1. 找设备端口（vendor-agnostic AI 助理端口）
        val port = DevicePortLocator.activeAssistantPort()
            ?: throw IllegalStateException("assistant.no_device: no active assistant port")
        devicePort = port

        // 2. 创建 chat agent 实例
        val a = NativeAgentRegistry.create(request.agentType)
        agent = a

        val sink = AssistantSinkAdapter(sessionId, emit)
        a.initialize(request.agentConfig, sink, context)

        // 3. chat agent connectService 立即上报 ready；保留 timeout 兜底以防未来变更
        a.connectService()
        try {
            withTimeout(CONNECT_TIMEOUT_MS) {
                sink.connected.await()
            }
        } catch (e: TimeoutCancellationException) {
            throw IllegalStateException(
                "assistant.connect_timeout: agent connect timeout after ${CONNECT_TIMEOUT_MS}ms"
            )
        }

        // 4. 协商外部音频格式：PCM_S16LE / 16k / mono / 20ms
        val format = ExternalAudioFormat.PCM_S16LE_16K_MONO_20MS
        val cap = a.externalAudioCapability()
        if (!cap.acceptsPcm) {
            throw IllegalStateException(
                "assistant.agent_unsupported: agent doesn't accept PCM external audio"
            )
        }

        // 5. 启动外部音频源 + 注入 TTS 反向 sink（回灌耳机扬声器）
        a.startExternalAudio(format, ttsSink(port))

        // 6. enter device port —— 此后耳机帧开始派发到 audioFrames flow
        port.enter(AssistantAudioFormat.PCM_S16LE_16K_MONO_20MS)

        // 7. 启动两条独立的 collect 协程
        scope.launch { pumpAudioFrames(port) }
        scope.launch { pumpDeviceErrors(port) }

        state = State.ACTIVE
        emit(AssistantEvents.sessionState(sessionId, "active"))
        Log.d(TAG, "session=$sessionId active")
    }

    private suspend fun pumpAudioFrames(port: DeviceAssistantPort) {
        try {
            var upCount = 0L
            var lastReportMs = System.currentTimeMillis()
            port.audioFrames.collect { frame ->
                upCount++
                val now = System.currentTimeMillis()
                if (now - lastReportMs >= 1000L) {
                    Log.d(TAG, "pump stats (last 1s): up=$upCount")
                    upCount = 0; lastReportMs = now
                }
                agent?.pushExternalAudioFrame(frame.bytes)
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "pumpAudioFrames error", e)
            emit(AssistantEvents.error(
                sessionId, "assistant.pump_failed", e.message ?: "unknown", fatal = true,
            ))
            stop()
        }
    }

    private suspend fun pumpDeviceErrors(port: DeviceAssistantPort) {
        port.errors.collect { err ->
            // device 错误默认非致命
            emit(AssistantEvents.error(sessionId, err.code, err.message))
        }
    }

    private fun ttsSink(port: DeviceAssistantPort) =
        object : ExternalAudioSink {
            override fun onTtsFrame(frame: ExternalAudioFrame) {
                if (frame.codec != ExternalAudioFormat.Codec.PCM_S16LE) {
                    Log.w(TAG, "tts frame codec ${frame.codec} not supported by device port; dropped")
                    return
                }
                runCatching {
                    port.reportPlayback(
                        AssistantPlaybackFrame(
                            codec = AssistantAudioCodec.PCM_S16LE,
                            sampleRate = frame.sampleRate,
                            channels = frame.channels,
                            bytes = frame.bytes,
                            isFinal = frame.isFinal,
                        )
                    )
                }.onFailure {
                    Log.w(TAG, "reportPlayback failed: ${it.message}")
                }
            }

            override fun onError(code: String, message: String) {
                emit(AssistantEvents.error(sessionId, code, message))
            }
        }

    /** 停止会话，释放所有资源。幂等。 */
    fun stop() {
        if (state == State.STOPPED || state == State.STOPPING) return
        state = State.STOPPING
        emit(AssistantEvents.sessionState(sessionId, "stopping"))
        Log.d(TAG, "session=$sessionId stopping")

        runCatching { devicePort?.exit() }
        runCatching { agent?.stopExternalAudio() }
        runCatching { agent?.disconnectService() }
        runCatching { agent?.release() }

        scope.cancel()
        agent = null
        devicePort = null

        state = State.STOPPED
        emit(AssistantEvents.sessionState(sessionId, "stopped"))
        Log.d(TAG, "session=$sessionId stopped")
    }

    fun markError(code: String, message: String) {
        if (state == State.STOPPED || state == State.ERROR) return
        emit(AssistantEvents.error(sessionId, code, message, fatal = true))
        emit(AssistantEvents.sessionState(sessionId, "error", message))
        runCatching { stop() }
        state = State.ERROR
    }
}

/** startAssistant 入参。 */
internal data class AssistantRequest(
    val agentType: String,
    val agentConfig: NativeAgentConfig,
    val userLanguage: String,
)
